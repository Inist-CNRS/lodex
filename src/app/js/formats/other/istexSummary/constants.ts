export const SORT_YEAR_DESC = 'YEAR_DESC';
export const SORT_YEAR_ASC = 'YEAR_ASC';

export const SORT_YEAR_VALUES = [SORT_YEAR_DESC, SORT_YEAR_ASC];

export const CUSTOM_ISTEX_QUERY = 'custom_istex_query';
export const HOST_ISSN = 'host.issn';
export const HOST_EISSN = 'host.eissn';
export const HOST_ISBN = 'host.isbn';
export const HOST_EISBN = 'host.eisbn';
export const HOST_TITLE = 'host.title';
export const HOST_TITLE_RAW = 'host.title.raw';
export const REFBIBS_HOST_TITLE_RAW = 'refBibs.host.title.raw';
export const REFBIBS_TITLE = 'refBibs.title';
export const TOP_HITS = 'topHits';

export const SEARCHED_FIELD_VALUES = [
    CUSTOM_ISTEX_QUERY,
    HOST_ISSN,
    HOST_EISSN,
    HOST_ISBN,
    HOST_EISBN,
    HOST_TITLE,
];
