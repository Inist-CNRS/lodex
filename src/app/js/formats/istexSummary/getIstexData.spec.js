import {
    getVolumeUrl,
    getIssueUrl,
    getDocumentUrl,
    getYearUrl,
    parseYearData,
    parseFacetData,
    getOtherVolumeUrl,
    getOtherIssueUrl,
    parseOtherData,
} from './getIstexData';

import { CUSTOM_ISTEX_QUERY } from './constants';

describe('getIstexData', () => {
    describe('parseFacetData', () => {
        it('should retrieve facet key for given facet name', () => {
            expect(
                parseFacetData('facet.name')({
                    response: {
                        aggregations: {
                            'facet.name': {
                                buckets: [
                                    { key: 'value1', docCount: 1 },
                                    { key: 'value2', docCount: 2 },
                                    { key: 'value3', docCount: 3 },
                                    { key: 'value4', docCount: 4 },
                                ],
                            },
                        },
                    },
                }),
            ).toEqual({
                hits: [
                    { name: 'value1', count: 1 },
                    { name: 'value2', count: 2 },
                    { name: 'value3', count: 3 },
                    { name: 'value4', count: 4 },
                ],
            });
        });
    });

    describe('getYearUrl', () => {
        it('should return url to get year facet', () => {
            expect(
                getYearUrl({
                    resource: {
                        value: 'issn',
                    },
                    field: {
                        name: 'value',
                    },
                    searchedField: 'searched.field',
                }),
            ).toMatchSnapshot();
        });

        it('should return the value if the custom ISTEX query is selected', () => {
            const query = 'CUSTOM_QUERY';

            const result = getYearUrl({
                resource: { value: query },
                field: { name: 'value' },
                searchedField: CUSTOM_ISTEX_QUERY,
            });

            expect(result).toMatchSnapshot();
        });
    });

    describe('parseYearData', () => {
        it('should extract year from data', () => {
            expect(
                parseYearData({
                    aggregations: {
                        publicationDate: {
                            buckets: [
                                { keyAsString: 2000, docCount: 1 },
                                { keyAsString: 2010, docCount: 2 },
                                { keyAsString: 2006, docCount: 3 },
                                { keyAsString: 2005, docCount: 4 },
                            ],
                        },
                    },
                }),
            ).toEqual({
                hits: [
                    { name: 2010, count: 2 },
                    { name: 2006, count: 3 },
                    { name: 2005, count: 4 },
                    { name: 2000, count: 1 },
                ],
            });
        });

        it('should extract year from data', () => {
            expect(
                parseYearData(
                    {
                        aggregations: {
                            publicationDate: {
                                buckets: [
                                    { keyAsString: 2000, docCount: 1 },
                                    { keyAsString: 2010, docCount: 2 },
                                    { keyAsString: 2006, docCount: 3 },
                                    { keyAsString: 2005, docCount: 4 },
                                ],
                            },
                        },
                    },
                    'YEAR_ASC',
                ),
            ).toEqual({
                hits: [
                    { name: 2000, count: 1 },
                    { name: 2005, count: 4 },
                    { name: 2006, count: 3 },
                    { name: 2010, count: 2 },
                ],
            });
        });

        it('should return empty array if empty response', () => {
            expect(parseYearData({})).toEqual({ hits: [] });
        });
    });

    describe('getVolumeUrl', () => {
        it('should return url to get volume facet', () => {
            expect(
                getVolumeUrl({
                    value: 'issn',
                    year: 'year',
                    searchedField: 'searched.field',
                })(),
            ).toMatchSnapshot();
        });
    });

    describe('getIssueUrl', () => {
        it('should return url to get issue facet', () => {
            expect(
                getIssueUrl({
                    value: 'issn',
                    year: 'year',
                    volume: 'volume',
                    searchedField: 'searched.field',
                })(),
            ).toMatchSnapshot();
        });

        it('should return url to get issue facet with other volume retrieving all non numerical volume', () => {
            expect(
                getIssueUrl({
                    value: 'issn',
                    year: 'year',
                    volume: 'other',
                    searchedField: 'searched.field',
                })(),
            ).toMatchSnapshot();
        });
    });

    describe('getDocumentUrl', () => {
        it('should return url to get document', () => {
            expect(
                getDocumentUrl({
                    value: 'issn',
                    year: 'year',
                    volume: 'volume',
                    issue: 'issue',
                    searchedField: 'searched.field',
                    documentSortBy: 'title.raw',
                })(),
            ).toMatchSnapshot();
        });

        it('should return url to get document with other volume retrieving all non numerical volume', () => {
            expect(
                getDocumentUrl({
                    value: 'issn',
                    year: 'year',
                    volume: 'other',
                    issue: 'issue',
                    searchedField: 'searched.field',
                    documentSortBy: 'title.raw',
                })(),
            ).toMatchSnapshot();
        });

        it('should return url to get document with other issue retrieving all non numerical issue', () => {
            expect(
                getDocumentUrl({
                    value: 'issn',
                    year: 'year',
                    volume: 'volume',
                    issue: 'other',
                    searchedField: 'searched.field',
                    documentSortBy: 'title.raw',
                })(),
            ).toMatchSnapshot();
        });
    });

    describe('getOtherVolumeUrl', () => {
        it('should return url to get non numerical volumes', () => {
            expect(
                getOtherVolumeUrl({
                    value: 'issn',
                    year: 'year',
                    volume: 'volume',
                    searchedField: 'searched.field',
                })(),
            ).toMatchSnapshot();
        });
    });

    describe('getOtherIssueUrl', () => {
        it('should return url to get non numerical issues', () => {
            expect(
                getOtherIssueUrl({
                    value: 'issn',
                    year: 'year',
                    volume: 'volume',
                    issue: 'issue',
                    searchedField: 'searched.field',
                })(),
            ).toMatchSnapshot();
        });
    });

    describe('parseOtherData', () => {
        it('should retrieve total of response and assign it to ther result', () => {
            expect(
                parseOtherData({
                    response: {
                        total: 19,
                    },
                }),
            ).toEqual({ name: 'other', count: 19 });
        });
    });
});
