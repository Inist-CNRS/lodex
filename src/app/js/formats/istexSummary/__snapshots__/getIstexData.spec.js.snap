// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIstexData getDocumentUrl should return url to get document 1`] = `
Object {
  "url": "https://api.istex.fr/document/?output=id%2CarkIstex%2Ctitle%2CpublicationDate%2Cauthor%2Chost.genre%2Chost.title&size=10&sortBy=title.raw&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22%20AND%20host.volume%3A%22volume%22%20AND%20host.issue%3A%22issue%22)",
}
`;

exports[`getIstexData getDocumentUrl should return url to get document with other issue retrieving all non numerical issue 1`] = `
Object {
  "url": "https://api.istex.fr/document/?output=id%2CarkIstex%2Ctitle%2CpublicationDate%2Cauthor%2Chost.genre%2Chost.title&size=10&sortBy=title.raw&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22%20AND%20host.volume%3A%22volume%22%20AND%20-host.issue%3A%5B0%20TO%20*%5D)",
}
`;

exports[`getIstexData getDocumentUrl should return url to get document with other volume retrieving all non numerical volume 1`] = `
Object {
  "url": "https://api.istex.fr/document/?output=id%2CarkIstex%2Ctitle%2CpublicationDate%2Cauthor%2Chost.genre%2Chost.title&size=10&sortBy=title.raw&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22%20AND%20-host.volume%3A%5B0%20TO%20*%5D%20AND%20host.issue%3A%22issue%22)",
}
`;

exports[`getIstexData getIssueUrl should return url to get issue facet 1`] = `
Object {
  "url": "https://api.istex.fr/document/?facet=host.issue%5B*-*%3A1%5D&output=*&size=0&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22%20AND%20host.volume%3A%22volume%22)",
}
`;

exports[`getIstexData getIssueUrl should return url to get issue facet with other volume retrieving all non numerical volume 1`] = `
Object {
  "url": "https://api.istex.fr/document/?facet=host.issue%5B*-*%3A1%5D&output=*&size=0&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22%20AND%20-host.volume%3A%5B0%20TO%20*%5D)",
}
`;

exports[`getIstexData getOtherIssueUrl should return url to get non numerical issues 1`] = `
Object {
  "url": "https://api.istex.fr/document/?output=*&size=0&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22%20AND%20host.volume%3A%22volume%22%20AND%20-host.issue%3A%5B0%20TO%20*%5D)",
}
`;

exports[`getIstexData getOtherVolumeUrl should return url to get non numerical volumes 1`] = `
Object {
  "url": "https://api.istex.fr/document/?output=*&size=0&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22%20AND%20-host.volume%3A%5B0%20TO%20*%5D)",
}
`;

exports[`getIstexData getVolumeUrl should return url to get volume facet 1`] = `
Object {
  "url": "https://api.istex.fr/document/?facet=host.volume%5B*-*%3A1%5D&output=*&size=0&q=(searched.field%3A%22issn%22%20AND%20publicationDate%3A%22year%22)",
}
`;

exports[`getIstexData getYearUrl should return the value if the custom ISTEX query is selected 1`] = `"https://api.istex.fr/document/?facet=publicationDate%5BperYear%5D&output=*&size=0&q=((CUSTOM_QUERY))"`;

exports[`getIstexData getYearUrl should return url to get year facet 1`] = `"https://api.istex.fr/document/?facet=publicationDate%5BperYear%5D&output=*&size=0&q=(searched.field%3A%22issn%22)"`;
